File: app/src/androidTest/java/com/bbtest/perform/base/FpsInfoBak.java
package com.bbtest.perform.base;

import android.util.Log;
import android.view.Choreographer;

import androidx.test.internal.util.ReflectionUtil;

import java.util.concurrent.atomic.AtomicInteger;

public class FpsInfoBak {

    private static FpsInfoBak INS = null;

    private AtomicInteger i = new AtomicInteger(0);
    private boolean pause = false;
    private static Thread m_t = null;

    private static int m_sm = 0;
    private static int m_fps = 0;
    private static int m_tryFpsCount = 0;
    private static boolean m_soLoaded = false;

    private static long mLastFrameTime = -1L;
    private static boolean mWarningFrameSkip = false;
    private static int Frame_Skip_Timeout = 100000000;

    private static Object mUserActionStatManager;

    public static boolean isWarningFrameSkip() {
        return mWarningFrameSkip;
    }

    public static void setWarningFrameSkip(boolean val) {
        mWarningFrameSkip = val;
    }

    public static int getCurrentSM() {
        return m_sm;
    }

    public static int getCurrentFPS() {
        return m_fps;
    }

    Choreographer.FrameCallback c = new Choreographer.FrameCallback() {

        public void doFrame(long frameTimeNanos) {
            if (FpsInfoBak.isWarningFrameSkip()) {
                warningFrameSkip(frameTimeNanos);
            } else {
                recordFps(frameTimeNanos);
            }
        }

        private void recordFps(long frameTimeNanos) {
            if (!FpsInfoBak.this.pause) {
                Choreographer.getInstance().postFrameCallback(this);
            }
            FpsInfoBak.this.i.incrementAndGet();
        }

        private void warningFrameSkip(long frameTimeNanos) {
            if (FpsInfoBak.this.pause) {
                return;
            }

            long eTime = System.nanoTime() - FpsInfoBak.mLastFrameTime;

            if ((eTime > FpsInfoBak.Frame_Skip_Timeout) && (eTime < 1000000000L)) {
                Log.e("QQDriverMonitor",
                        "！！！！掉帧！！！etime = " + eTime / 1000000L);
                if (FpsInfoBak.mUserActionStatManager == null) {
//                    FpsInfoBak.mUserActionStatManager = ReflectionUtil.invokeStaticMethod("com.tencent.mtt.base.stat.UserActionStatManager", "getInstance", null);
                }
                if (FpsInfoBak.mUserActionStatManager != null) {
//                    Log.e("QQDriverMonitor", "user action " + ReflectionUtil.invokeMethod(FpsInfoBak.mUserActionStatManager, "generateUserAcList", new Object[]{Boolean.valueOf(true), Boolean.valueOf(true), Boolean.valueOf(false)}));
                }
            }

            FpsInfoBak.mLastFrameTime = frameTimeNanos;
            Choreographer.getInstance().postFrameCallback(this);
        }
    };

    Runnable r = new Runnable() {

        public void run() {

            while (!FpsInfoBak.this.pause) {

                FpsInfoBak.m_sm = FpsInfoBak.this.i.getAndSet(0);
//                FpsInfoBak.m_fps = FpsInfoBak.access$8();
                try {
                    Thread.sleep(1000L);
                } catch (InterruptedException localInterruptedException) {
                }
            }
        }
    };


    public static FpsInfoBak getIns() {
        if (INS == null) {
            INS = new FpsInfoBak();
        }
        return INS;
    }


    public void CallinOnCreate() {
        if (m_t == null) {
            this.pause = false;

            Choreographer.getInstance().postFrameCallback(this.c);

            m_t = new Thread(this.r);
            m_t.start();
            m_t.setName("SM calculator");
        }
    }


    public void CallinOnPause() {
        this.pause = true;
    }

    private static int getFPSIfPossible() {
        int fps = 0;
        if (m_soLoaded) {
            try {
                fps = getFPS();
            } catch (Throwable localThrowable) {
            }
        }
        return fps;
    }

    public static void setSOLoaded(boolean loaded) {
        m_soLoaded = loaded;
    }

    public static native int getFPS();
}

File: app/src/androidTest/kotlin/com/bbtest/perform/base/FpsInfoBak.kt
package com.bbtest.perform.base

import android.util.Log
import android.view.Choreographer
import java.util.concurrent.atomic.AtomicInteger

class FpsInfoBak {
    private val i = AtomicInteger(0)
    private var pause = false

    private val frameCallback = object : Choreographer.FrameCallback {
        override fun doFrame(frameTimeNanos: Long) {
            if (isWarningFrameSkip()) {
                warningFrameSkip(frameTimeNanos)
            } else {
                recordFps()
            }
        }

        private fun recordFps() {
            if (!pause) {
                Choreographer.getInstance().postFrameCallback(this)
            }
            i.incrementAndGet()
        }

        private fun warningFrameSkip(frameTimeNanos: Long) {
            if (pause) {
                return
            }

            val eTime = System.nanoTime() - mLastFrameTime
            if (eTime > frameSkipTimeout && eTime < 1000000000L) {
                Log.e("QQDriverMonitor", "！！！！掉帧！！！etime = ${eTime / 1000000L}")
            }

            mLastFrameTime = frameTimeNanos
            Choreographer.getInstance().postFrameCallback(this)
        }
    }

    private val runnable = Runnable {
        while (!pause) {
            m_sm = i.getAndSet(0)
            try {
                Thread.sleep(1000L)
            } catch (_: InterruptedException) {
            }
        }
    }

    fun CallinOnCreate() {
        if (m_t == null) {
            pause = false
            Choreographer.getInstance().postFrameCallback(frameCallback)
            m_t = Thread(runnable)
            m_t?.start()
            m_t?.name = "SM calculator"
        }
    }

    fun CallinOnPause() {
        pause = true
    }

    companion object {
        private var INS: FpsInfoBak? = null
        private var m_t: Thread? = null
        private var m_sm = 0
        private var m_fps = 0
        private var m_soLoaded = false
        private var mLastFrameTime = -1L
        private var mWarningFrameSkip = false
        private var frameSkipTimeout = 100000000

        @JvmStatic
        fun isWarningFrameSkip(): Boolean = mWarningFrameSkip

        @JvmStatic
        fun setWarningFrameSkip(value: Boolean) {
            mWarningFrameSkip = value
        }

        @JvmStatic
        fun getCurrentSM(): Int = m_sm

        @JvmStatic
        fun getCurrentFPS(): Int = m_fps

        @JvmStatic
        fun getIns(): FpsInfoBak {
            if (INS == null) {
                INS = FpsInfoBak()
            }
            return requireNotNull(INS)
        }

        private fun getFPSIfPossible(): Int {
            var fps = 0
            if (m_soLoaded) {
                try {
                    fps = getFPS()
                } catch (_: Throwable) {
                }
            }
            return fps
        }

        @JvmStatic
        fun setSOLoaded(loaded: Boolean) {
            m_soLoaded = loaded
        }

        @JvmStatic
        external fun getFPS(): Int
    }
}
